core/src/main/java/io/jenkins/pluginhealth/scoring/probes/IncrementalBuildDetectionProbe.java

```diff
@@ -0,0 +1,147 @@
+ * SOFTWARE.
+ */
+
+package io.jenkins.pluginhealth.scoring.probes;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+
+import io.jenkins.pluginhealth.scoring.model.Plugin;
+import io.jenkins.pluginhealth.scoring.model.ProbeResult;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.annotation.Order;
+import org.springframework.stereotype.Component;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.xml.sax.SAXException;
+
+/**
+ * This probe checks for Incremental Build configuration in plugins.
+ * It looks for configuration in {@code .mvn/maven.config} and {@code .mvn/extensions.xml} files.
+ * The probe is successful when the configurations are found in both the files. Otherwise, it fails.
+ */
+@Component
+@Order(IncrementalBuildDetectionProbe.ORDER)
+public class IncrementalBuildDetectionProbe extends Probe {
+    public static final int ORDER = SCMLinkValidationProbe.ORDER + 100;
+    public static final String KEY = "incremental-build-maven-configuration";
+    private static final Logger LOGGER = LoggerFactory.getLogger(IncrementalBuildDetectionProbe.class);
+    private static final String INCREMENTAL_TOOL = "io.jenkins.tools.incrementals";
+    private static final String INCREMENTAL_TOOL_ARTIFACT_ID = "git-changelist-maven-extension";
+
+    @Override
+    protected ProbeResult doApply(Plugin plugin, ProbeContext context) {
+        final Path scmRepository = context.getScmRepository();
+        final Path mvnConfig = scmRepository.resolve(".mvn");
+        if (Files.notExists(mvnConfig)) {
+            LOGGER.info("Could not find Maven configuration folder {} plugin while running {} probe.", plugin.getName(), key());
+            return ProbeResult.failure(key(), String.format("Could not find Maven configuration folder for the %s plugin.", plugin.getName()));
+        }
+
+        Supplier<Stream<Path>> incrementalConfigs = () -> {
+            try {
+                return Files.find(mvnConfig, 1, (path, $) -> Files.isRegularFile(path) && (path.endsWith("maven.config") || path.endsWith("extensions.xml")));
+            } catch (IOException e) {
+                LOGGER.error("Could not read files from .mvn directory for {} plugin while running {} probe.", plugin.getName(), key());
+                return null;
+            }
+        };
+        Optional<Path> mavenExtensionsFile = incrementalConfigs.get().filter(path -> path.endsWith("extensions.xml")).findFirst();
+        Optional<Path> mavenConfigFile = incrementalConfigs.get().filter(path -> path.endsWith("maven.config")).findFirst();
```

Comment:
Your code works, but I just wanted to share something here to ensure you understand the whole thing.

In that particular case, it is not a problem, but basically, that piece of code will explore the `mvnConfig` directory each time you will call the supplier. Not a big deal in our case because we do not expect that directory to contain a lot of files, but in other similar situations, this could be an issue, because there is no particular need of actually processing *x* times the same folder.

The problem you are trying to solve here is that a `Stream` can only be consumed once, and the Supplier approach allows you to retrieve a new `Stream` everytime you call the supplier. But you can actually implement a `Supplier` that will create for you a new `Stream` out of data which you already collected.

See for example:

```java
try (Stream<Path> incrementalConfigsStream = Files.find(mvnConfig, 1, (path, $) -> Files.isRegularFile(path) && (path.endsWith("maven.config") || path.endsWith("extensions.xml")))) {
    List<Path> incrementalConfigsStore = incrementalConfigsStream.toList();
    
    // Here, we stored the data in a `List`, and we can create a supplier that will create a new `Stream` for us from this data
    Supplier<Stream<String>> incrementalConfigs = () -> incrementalConfigsStore.stream();

    // And now you can use your `Supplier` to retrieve whatever you want, and the `mvnConfig` directory will be read just once
        Optional<Path> mavenExtensionsFile = incrementalConfigs.get().filter(path -> path.endsWith("extensions.xml")).findFirst();
        Optional<Path> mavenConfigFile = incrementalConfigs.get().filter(path -> path.endsWith("maven.config")).findFirst();

}
```

WDYT?

Reply:
This is a much better of doing things. 

I do not completely understand `Functional Interface` and  `Supplier` that well. They still scare me.

Reply:
These are common things coming from Functional Programming. The concepts apply to lots of languages, not Java only.

You can have a look at this 101 guide from Cassidy Williams for example: https://github.com/readme/guides/functional-programming-basics

If you're still unsure and would like to discuss it a bit so we can share some examples live, LMK, we can have a quick call for more explanation.

Reply:
Thanks for the link and offering for a quick meeting. I will let you know if the need arises.